pymmcore_plus/core/_mmcore_plus.py

```diff
@@ -111,38 +112,25 @@ def __del__(self):
     def setProperty(
         self, label: str, propName: str, propValue: Union[bool, float, int, str]
     ) -> None:
-        """setProperty with more reliable event emission.
-
-        As stated by Nico: "Callbacks are mainly used to give devices the opportunity to
-        signal back to the UI."
-        https://forum.image.sc/t/micromanager-events-core-events-not-coming-through/53014/2
-
-        Because it's left to the device adapter to emit a signal, in many cases uses
-        `setProperty()` will NOT lead to a new `propertyChanged` event getting emitted.
-        But that makes it hard to create listeners (i.e. in the gui or elsewhere).
+        """setProperty with reliable event emission."""
+        with self._property_change_emission_ensured(label, (propName,)):
+            super().setProperty(label, propName, propValue)
 
-        While this method override cannot completely solve that problem (core-internal
-        changes will still lack an associated event emission in many cases), it can at
-        least guarantee that if we use `CMMCorePlus.setProperty` to change the property,
-        then a `propertyChanged` event will be emitted if the value did indeed change.
+    @synchronized(lock)
+    def setState(self, stateDeviceLabel: str, state: int) -> None:
+        """Set state (by position) on stateDeviceLabel, with reliable event emission."""
+        with self._property_change_emission_ensured(
+            stateDeviceLabel, ("State", "Label")
+        ):
+            super().setState(stateDeviceLabel, state)
 
-        Parameters
-        ----------
-        label : str
-            device label
-        propName : str
-            property name
-        propValue : Union[bool, float, int, str]
-            new value
-        """
-        before = super().getProperty(label, propName)
-        with _blockSignal(
-            self.events, self.events.propertyChanged
-        ):  # block the native event.
-            super().setProperty(label, propName, propValue)
-        after = super().getProperty(label, propName)
-        if before != after:
-            self.events.propertyChanged.emit(label, propName, after)
+    @synchronized(lock)
+    def setStateLabel(self, stateDeviceLabel: str, stateLabel: str) -> None:
+        """Set state (by label) on stateDeviceLabel, with reliable event emission."""
+        with self._property_change_emission_ensured(
+            stateDeviceLabel, ("State", "Label")

```

Comment:
What happens if a state doesn't have an associated label? I suspect that a blank string will be sent, but we should check that this doesn't cause an error before merging.

Reply:
Looks like they they default to `State-1` and all is well